### ISSUE TITLE
"Bad item skipped" / "Invalid item" when adding event to shared calendar (O365, Outlook.com) -> bad default value for privacy

### ISSUE THREAD

**Issue by DerBachmannRocker** [jobisoft/TbSync#74]
TbSync version: 0.7.8
Thunderbird version: 52.7.0

Scenario: I use Thunderbird + Lightning + TbSync to view and edit multiple calendars on Outlook.com. I create a new calendar item in a remote calendar that was shared with me. I have write permission on this calendar, but I do not own it.

I then hit the sync button to write to Outlook.com and get the following in the sync status window:
The EAS server rejected your items for the following reasons: 1x <invalid XML (possible bug in TbSync)>. 

Solution: there is no default-value for privacy set. If I open the offending calendar item, click once on the "privacy" button (so "public" is set), close and sync again, everything works as expected.

I looked around but was unable to find a default value setting for Lightning regarding privacy. It would be nice if TbSync could enforce a default value if none is set (presumably "public" should be correct in most cases).

**Comment by jobisoft** [comment]
Thanks for the detailed report. However, this is strange. I DO use a default value for the privacy field. This is how I map the Thunderbird values to EAS values:

'''
        //EAS Sensitivity :  0 = Normal  |  1 = Personal  |  2 = Private  |  3 = Confidential
        Sensitivity : { "0":"PUBLIC", "1":"unset", "2":"PRIVATE", "3":"CONFIDENTIAL"}, //to CLASS
'''

Which means, if none is selected, I send a EAS sensitivity of 1 = Personal. You can check that by opening the debug log after you synced a new event from TB to your server. It wil include something like this:

'''
&#x3C;ApplicationData&#x3E;
&#x3C;TimeZone xmlns=&#x27;Calendar&#x27;&#x3E;xP%2F%2F%2F0UAdQByAG8AcABlAC8AQgBlAHIAbABpAG4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAoAAAAFAAIAAAAAAAAAAAAAAEUAdQByAG8AcABlAC8AQgBlAHIAbABpAG4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAAAFAAEAAAAAAAAAxP%2F%2F%2Fw%3D%3D&#x3C;/TimeZone&#x3E;
&#x3C;AllDayEvent xmlns=&#x27;Calendar&#x27;&#x3E;0&#x3C;/AllDayEvent&#x3E;
&#x3C;Body xmlns=&#x27;AirSyncBase&#x27;&#x3E;
&#x3C;Type xmlns=&#x27;AirSyncBase&#x27;&#x3E;1&#x3C;/Type&#x3E;
&#x3C;EstimatedDataSize xmlns=&#x27;AirSyncBase&#x27;&#x3E;4&#x3C;/EstimatedDataSize&#x3E;
&#x3C;Data xmlns=&#x27;AirSyncBase&#x27;&#x3E;Test&#x3C;/Data&#x3E;
&#x3C;/Body&#x3E;
&#x3C;BusyStatus xmlns=&#x27;Calendar&#x27;&#x3E;2&#x3C;/BusyStatus&#x3E;
&#x3C;DtStamp xmlns=&#x27;Calendar&#x27;&#x3E;20180509T212841Z&#x3C;/DtStamp&#x3E;
&#x3C;EndTime xmlns=&#x27;Calendar&#x27;&#x3E;20180515T220000Z&#x3C;/EndTime&#x3E;
&#x3C;Location xmlns=&#x27;Calendar&#x27;/&#x3E;
&#x3C;Sensitivity xmlns=&#x27;Calendar&#x27;&#x3E;1&#x3C;/Sensitivity&#x3E;
&#x3C;Subject xmlns=&#x27;Calendar&#x27;&#x3E;Neuer%20Termin&#x3C;/Subject&#x3E;
&#x3C;StartTime xmlns=&#x27;Calendar&#x27;&#x3E;20180515T210000Z&#x3C;/StartTime&#x3E;
&#x3C;UID xmlns=&#x27;Calendar&#x27;&#x3E;ffd3433d-281c-4270-821b-3bb2e5bfee28&#x3C;/UID&#x3E;
&#x3C;MeetingStatus xmlns=&#x27;Calendar&#x27;&#x3E;1&#x3C;/MeetingStatus&#x3E;
&#x3C;Attendees xmlns=&#x27;Calendar&#x27;/&#x3E;
&#x3C;Categories xmlns=&#x27;Calendar&#x27;/&#x3E;
&#x3C;/ApplicationData&#x3E;
'''

So, what I conclude now, outlook.com does not allow "personal" in a shared calendar. Which is a protocol violation. I may forward this to Microsoft, because I do not think this is a bug in TbSync.

Do you have any more details on this?

**Comment by SlimG** [comment]
I struggle with the same problem, and I can work around it by setting privacy in my new events to any privacy.

Looking trough the TbSync log it looks like there is no lines starting with "CLASS" in the events that won't sync to O365.

Here is my offending event:

> BEGIN:VCALENDAR
> PRODID:-//Mozilla.org/NONSGML Mozilla Calendar V1.1//EN
> VERSION:2.0
> BEGIN:VTIMEZONE
> TZID:Europe/Oslo
> BEGIN:DAYLIGHT
> TZOFFSETFROM:+0100
> TZOFFSETTO:+0200
> TZNAME:CEST
> DTSTART:19700329T020000
> RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3
> END:DAYLIGHT
> BEGIN:STANDARD
> TZOFFSETFROM:+0200
> TZOFFSETTO:+0100
> TZNAME:CET
> DTSTART:19701025T030000
> RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10
> END:STANDARD
> END:VTIMEZONE
> BEGIN:VEVENT
> CREATED:20181023T143954Z
> LAST-MODIFIED:20181023T144038Z
> DTSTAMP:20181023T144038Z
> UID:8abc7515-c11e-43b8-894d-d32b0c9d2a14
> SUMMARY:This does not sync to O365
> DTSTART;TZID=Europe/Oslo:20181023T170000
> DTEND;TZID=Europe/Oslo:20181023T173000
> TRANSP:OPAQUE
> END:VEVENT
> END:VCALENDAR

**Comment by SlimG** [comment]
The same event is sent to O365 when I manually set the privacy to Public
Here is the event as it looks when exported from Thunderbird:

> BEGIN:VEVENT
> CREATED:20181023T143954Z
> LAST-MODIFIED:20181023T144524Z
> DTSTAMP:20181023T144524Z
> UID:nY_WbxgSW0ayTIdCiVIooQABoOO5Ap2P1m8YEltGskyHQolSKKEAAaDlANI1
> SUMMARY:This does not sync to O365
> DTSTART;TZID=Europe/Oslo:20181023T170000
> DTEND;TZID=Europe/Oslo:20181023T173000
> TRANSP:OPAQUE
> CLASS:PUBLIC
> END:VEVENT

**Comment by SlimG** [comment]
Here I've created the same event again in Thunderbird without manually editing privacy:

> BEGIN:VEVENT
> CREATED:20181023T144911Z
> LAST-MODIFIED:20181023T144913Z
> DTSTAMP:20181023T144913Z
> UID:408e7060-ad77-4788-976c-716c44d724f0
> SUMMARY:This does not sync to O365
> DTSTART;TZID=Europe/Oslo:20181023T170000
> DTEND;TZID=Europe/Oslo:20181023T173000
> TRANSP:OPAQUE
> END:VEVENT

**Comment by jobisoft** [comment]
Thanks, the crucial information missing from you is, if the calendar you are using and having this issue with is also a shared calendar as with @DerBachmannRocker or if it is a normal calendar.

**Comment by SlimG** [comment]
The calendar in question is a calendar shared with me. Its not my own calendar. I frankly hav not tested my own calendar yet as it was most important to me that the shared one worked

**Comment by SlimG** [comment]
How about creating a configurable option with a dropdown list in TbSync?

### Workaround when privacy setting is missing
- Do not apply workaround (Default)
- Set Privacy to Normal
- Set Privacy to Personal
- Set Privacy to Private
- Set Privacy to Confidential

**Comment by jobisoft** [comment]
I investigated thus further. First of all, I added a web link to the error message, which will provide more details and how to get around this bug:

![image](https://user-images.githubusercontent.com/5830621/47460479-e2a55e00-d7de-11e8-861f-e8dee293bd40.png)

German translation, but you get the point...

Then I searched throu the sharing option in O365 and the person who shares the calendar has the option to share it in "secretary" mode, where I can activate the "private" mode.

![image](https://user-images.githubusercontent.com/5830621/47460610-2c8e4400-d7df-11e8-95a5-2fe0fc3ae0ec.png)

If that private flag is set, the error does not show up. If you look at my mapping:

'''
//EAS Sensitivity : 0 = Normal | 1 = Personal | 2 = Private | 3 = Confidential
Sensitivity : { "0":"PUBLIC", "1":"unset", "2":"PRIVATE", "3":"CONFIDENTIAL"}, //to CLASS, TB Privacy
'''

you see, that an unset privacy setting of Thunderbird maps to "EAS Personal" which corresponds to the private mode in O365. So this all looks "right" from O365. Just the message thrown by O365 could have been a better one (access denied or whatever). If the owner of the calendar does not give you the right to manage his private events, you cannot add such an event.

Now for convenience one could argue to change the default value of that Privacy setting to public, but that is not something I want to do, because the consequences of that would be more severe.

**I am sorry guys, this is a WONTFIX, you just have to select a privacy setting if you want to add events to a shared calendar.**
